Allow Ax commands to trigger workflow methods

danilvpetrov opened the pull request:
This PR is to allow triggering WF methods with Ax commands as a addition to the existing event triggering.

jmalloc commented on src/ax/saga/workflow.go:
```diff
@@ -22,30 +22,33 @@ type Workflow struct {
-// to be executed.
+// producing new commands, based on the command or event being handled. They may
+// inspect the current state of the workflow, and then return zero or more
+// commands to be executed.
```
Could you change this paragraph to simply read:

> Handler methods are responsible for mutating the state of the workflow and producing new commands, based on the message being handled.

The second sentence seems a bit redundant, not really related to your PR, but may as well fix it up since you're changing it already.

jmalloc commented on src/ax/saga/workflow.go:
```diff
@@ -22,30 +22,33 @@ type Workflow struct {
+//     func (*BankTransferWorkflow) StartWhenTransferStarted(*messages.StartTransfer, ax.CommandExecutor)
 //     func (*BankTransferWorkflow) StartWhenTransferStarted(*messages.TransferStarted, ax.CommandExecutor)
+//     func (*BankTransferWorkflow) WhenAccountDebited(*messages.DebitAccount, ax.CommandExecutor)
 //     func (*BankTransferWorkflow) WhenAccountDebited(*messages.AccountDebited, ax.CommandExecutor)
```
Hrm, I think these prefixes need to be revised.  So far throughout Ax, the word "when" is handler methods always implies an event.  In `Aggregate` for example, "event appliers" start with `When`, but command handlers do not require any specific naming.

We could keep `StartWhen` and `When` for events, and come up with some other prefixes for commands, but I'm not sure what they might be.  Any thoughts?

danilvpetrov replied:
Perhaps, it's still worth to differentiate signatures by command and event and not to use any prefixes for command signatures similar to aggregates?

jmalloc replied:
Yeah I think that might be best.  Aggregates don't currently require a prefix for commands, but I'm happy to add one there to keep consistency between aggregates and workflows.

/cc @koden-km

danilvpetrov commented:
Added more changes as per yesterday's discussion to modify prefixes for command and event handlers both for workflows and aggregates. Modified comments and docs to reflect the new changes,  changed banking example to work with new prefixes, tested banking example: everything is working fine. 

The summary of changes are as follows:

### 1. Aggregates 
Command handlers now must start with **'Do'**
As before,  all event handlers must start with **'When'**.

### 1. Workflows
Workflow-triggering handlers now must start with **'Begin'**. This works both for commands and events.
Non-triggering command handlers now must start with **'Do'**
As before, all event handlers must start with **'When'**; however, for workflow-triggering event handlers a combined prefix **'BeginWhen'** must be added.

jmalloc commented on src/ax/saga/workflow.go:
```diff
@@ -14,75 +15,122 @@ type Workflow struct {
+	Prototype     Data
+	Triggers      ax.MessageTypeSet
+	NonTriggers   ax.MessageTypeSet
+	CommandHandle typeswitch.Switch
```
These names should probs be `HandleCommand` and `HandleEvent`, for maximum English overdrive.

jmalloc commented on src/ax/saga/workflow.go:
```diff
@@ -14,75 +15,122 @@ type Workflow struct {
+// methods must have prefix "Do". Both triggering and non-triggering event
+// handler methods must have prefix "When". For triggering event handlers the
+// combined prefix results in "BeginWhen". By convention these prefixes are
+// followed by the message name, such as:
```
This section is a bit hard to explain, I had a go at rewording it:

> The names of handler methods are meaningful to the workflow system. If a message is meant to trigger a new workflow instance, its handler method's name must prefixed with "Begin", if it is a command handler, or "BeginWhen" if it is an event handler. Messages that can be routed to existing workflow instances, but not cause new instances must have their method names prefixed with "Do" and
"When" for commands and events, respectively. 
>
> By convention these prefixes are followed by the message name, such as:

jmalloc commented on src/ax/saga/workflow.go:
```diff
@@ -14,75 +15,122 @@ type Workflow struct {
 	}
 
+	// setup type-switch for command handlers.
+	csw, ctypes, err := typeswitch.New(
```
I can't remember... was there a reason we could 't use a single `typeswitch.New()` call with all of the signatures for both commands and events?

If not, we should do this, because it improves two things:

- removes the command/event `switch` from `HandleMessage()`
- prevents duplicate handlers for messages that are both events and commands (which is discouraged but allowed)

danilvpetrov replied:
As discussed, using two separate switches for commands and events is going to stay for now until a future version of switch (which is going to be a separate library) supports adding multiple cases to a single switch instance.
